libazureinit/src/goalstate.rs

```diff
@@ -41,34 +45,63 @@ pub struct RoleInstance {
     instance_id: String,
 }
 
-pub async fn get_goalstate(client: &Client) -> Result<Goalstate, Error> {
-    let url = "http://168.63.129.16/machine/?comp=goalstate";
+const DEFAULT_GOALSTATE_URL: &str =
+    "http://168.63.129.16/machine/?comp=goalstate";
+
+pub async fn get_goalstate(
+    client: &Client,
+    retry_interval: Duration,
+    total_timeout: Duration,
+    url: Option<&str>,
+) -> Result<Goalstate, Error> {
+    let url = url.unwrap_or(DEFAULT_GOALSTATE_URL);
 
     let mut headers = HeaderMap::new();
     headers.insert("x-ms-agent-name", HeaderValue::from_static("azure-init"));
     headers.insert("x-ms-version", HeaderValue::from_static("2012-11-30"));
 
-    let request = client.get(url).headers(headers);
-    let response = request.send().await?;
+    let response = timeout(total_timeout, async {
+        loop {
+            if let Ok(response) = client
+                .get(url)
+                .headers(headers.clone())
+                .timeout(Duration::from_secs(30))
+                .send()
+                .await
+            {
+                let statuscode = response.status();
 
-    if response.status().is_success() {
-        let body = response.text().await?;
+                if statuscode.is_success() && statuscode == StatusCode::OK {
+                    return Ok(response);
+                }
 
-        let goalstate: Goalstate = from_str(&body)?;
-        Ok(goalstate)
-    } else {
-        Err(Error::HttpStatus {
-            endpoint: url.to_owned(),
-            status: response.status(),
-        })
-    }
+                if http::RETRY_CODES.contains(&statuscode) {
```

Comment:
I think you meant to have 
```suggestion
                if !http::RETRY_CODES.contains(&statuscode) {
```

Since this didn't get caught by the tests, I think we'll need to adjust them a bit to make sure we can assert retry codes do attempt several times. Maybe we just turn the retry interval and total wait time down to something acceptable for tests (100ms and 3 seconds respectively? lower even?) and not artificially skip time ahead in some or all tests.

Reply:
Fixed the code as suggested, still some to do left.